Delay expansion errors for rules

rgrinberg opened the pull request:
Previosuly, a failed expansion would immediately fail generating a rule.
Now, we delay the error using a [Build.fail].

I've left some inline comments to make it easy for reviewers.

rgrinberg committed:
Add reproduction case for #3252

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
Delay expansion errors for rules

Previosuly, a failed expansion would immediately fail generating a rule.
Now, we delay the error using a [Build.fail].

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -245,14 +262,17 @@ module Resolved_forms = struct
+  let prefix_failures t b =
+    match t.failures with
+    | [] -> b
+    | fail :: _ -> Build.O.( >>> ) (Build.fail fail) b
```
This refactoring is useful to make sure that we use the same logic to prefix rules with the failure discovered in the expansion

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -322,11 +342,6 @@ let expand_and_record acc ~map_exe ~dep_kind ~expansion_kind
   in
-  let add_fail =
-    match expansion_kind with
-    | Static -> fun _ (f : fail) -> f.fail ()
```
This clause was the cause of the error. It's too soon to raise the error at this stage, we should instead prefix the rule itself with the error.

diml replied:
Makes sense

rgrinberg commented on src/dune/super_context.ml:
```diff
@@ -654,16 +679,31 @@ module Deps = struct
+                (fun () ->
+                  let loc = String_with_vars.loc sw in
+                  User_error.raise ~loc
+                    [ Pp.text "failed to expand percent form" ])
```
The user should only see this error if they used some unknown percent form. Because the errors discovered before the expansion are prefixed.

diml commented on src/dune/super_context.ml:
```diff
@@ -602,48 +602,73 @@ module Deps = struct
+      Expander.Partial.expand_path expander s
+      |> String_with_vars.Partial.map ~f:(fun path ->
+             let+ () = Build.path path in
+             [ path ])
```
I don't understand why we need to go through partial expansion here. The aim of partial expansion for actions is to be able to collect targets before the action is fully expanded. However, we never need this for the `deps` field given that it doesn't contain targets.

rgrinberg replied:
We need to interpret `deps` before expanding targets. For example, consider how we evaluate something like this:

```
(action
 (deps (:foo %{bin:bar}))
 (target %{foo}))
```

diml replied:
Agreed, but I still don't follow why we need to go through partial.

rgrinberg replied:
The errors are discovered as we're doing the expansion. When such an error is discovered, the expansion of that variable is aborted and hence the only way to recover is through a partial expansion.

I suppose it's possible to make this error some other exception. It could be caught and handled instead of handling partial expansions.

diml replied:
Ok, I feel like I'll need to go over all this code again at some point. I'm finding it difficult to follow everything. When you have left the airport chaos, we can maybe go over this together via videoconf.

Though, if someone else can review this don't wait for me.

rgrinberg replied:
In case someone else is curious, here's the pseudo-code of the current code works:

```
String_with_vars.expand ~f:(fun var ->
  if binary_vailable var then
    add_failure resolved_forms;
    None (* This raises because we are in an [expand] *)
  else
    Some (binary_path)
)
```

My proposal is to change the above to use `partial_expand` over `expand`. This will prevent the `None` case from raising and we'll be able to extract the error from the resolved form.

To avoid partial expansion, I suggested changing `add_failure resolved_forms` to raise an exception. The caller can catch it and propagate it via `Build.fail`.

diml replied:
Oh ok, I get it now!

In other places where we want to delay error reporting, such as for library name resolution, we are using the `Or_exn.t` monad. What do you think of doing the same here? That would make the intent explicit.

I was finding the code confusing because in my head partial is related to static vs dynamic. While here we are talking about delaying error reporting.

rgrinberg replied:
Sure, I can do that. It would also make things a bit faster as we'd stop trying to expand the other variables.

rgrinberg commented:
@diml should be ready now.

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -324,8 +351,9 @@ let expand_and_record acc ~map_exe ~dep_kind ~expansion_kind
     match expansion_kind with
-    | Static -> fun _ (f : fail) -> f.fail ()
-    | Dynamic -> Resolved_forms.add_fail
+    | Static -> Or_exn.raise
```
This wrapping is necessary so that the exception is caught by `Expander.Or_error`. We want to make sure that we don't accidentally delay exceptions that weren't raised by `add_fail`.

diml replied:
I'm wondering if that's really necessary. I tried removing the wrapping and nothing broke in the testsuite. I'm also thinking that we could remove `add_fail` entirely and just add a catch all around `expand_and_record`. That would make the code simpler and easier to reason about. WDYT?

rgrinberg replied:
> I'm wondering if that's really necessary. I tried removing the wrapping and nothing broke in the testsuite.

Indeed. Another option is to just catch & delay `User_error.E`. I think we can reasonably assume that all errors can be delayed.

>  I'm also thinking that we could remove add_fail entirely and just add a catch all around expand_and_record. That would make the code simpler and easier to reason about. WDYT?

Seems fine to me. Just to confirm, you would then have to handle this expansion based on the `expansion_kind`. I.e.

```ocaml
...
with User_error.E _ as exn ->
  match exception_kind with
  | Dynamic -> Resolved_forms.add acc exn
  | Static -> reraise exn
```

diml replied:
Yep. But I don't even think we need to match on `User_error.E` here. Catching all exceptions seems fine. We are just making the behaviour lazy, that's all.

rgrinberg replied:
I think it would be quite bad to swallow `Code_error` here. See the last commit, I've simplified the code further but I'm still not catching only User_errors. I think that's a safe compromise.

diml replied:
Seems reasonable

rgrinberg commented on src/dune/expander.ml:
```diff
@@ -409,24 +437,22 @@ let expand_and_record acc ~map_exe ~dep_kind ~expansion_kind
+        | false ->
+          if Lib.DB.available (Scope.libs t.scope) lib then
+            User_error.E
+              (User_error.make ~loc
```
This error is now created more eagerly than necessary. I don't think this is worth worrying about however.

jeremiedimino committed:
Simplify expansion code

In expand_and_record, instead of receiving an argument describing how
the function should behave, we always do the same thing and add two
wrappers that decide what to do based on what behavior the caller
wanted.

Make [Resolved_forms] hidden inside [Expander]

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

diml commented:
I did a pass of refactoring and made the API of `Expander` a bit higher level. For instance, `Resolved_forms` is no longer exposed and is just an implementation detail. I feel like `expander.ml` could be simplified further, but that'll be for another time.

rgrinberg commented:
@diml I removed the type variable from expander_result. Was it added by accident?

 I agree this looks much better.

diml commented:
> @diml I removed the type variable from expander_result. Was it added by accident?

Indeed. It was a temporary state while I was rediscovering how all this works.
